sdks/python/apache_beam/utils/interactive_utils.py

```diff
@@ -0,0 +1,98 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Common interactive utility module.
+
+For experimental usage only; no backwards-compatibility guarantees.
+"""
+from __future__ import absolute_import
+
+import logging
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def is_in_ipython():
+  """Determines if current code is executed within an ipython session."""
+  is_in_ipython = False
+  # Check if the runtime is within an interactive environment, i.e., ipython.
+  try:
+    from IPython import get_ipython  # pylint: disable=import-error
+    if get_ipython():
+      is_in_ipython = True
+  except ImportError:
+    pass  # If dependencies are not available, then not interactive for sure.
+  return is_in_ipython
+
+
+def is_in_notebook():
+  """Determines if current code is executed from an ipython notebook.
+
+  If is_in_notebook() is True, then is_in_ipython() must also be True.
+  """
+  is_in_notebook = False
+  if is_in_ipython():
+    # The import and usage must be valid under the execution path.
+    from IPython import get_ipython
+    if 'IPKernelApp' in get_ipython().config:
+      is_in_notebook = True
+  return is_in_notebook
+
+
+def alter_label_if_interactive(transform, pvalueish):
```

Comment:
This is more like alter_label_if_ipython. Do we want/need it to be more generic? Like supporting users running in a python shell in an interactive fashion?

Reply:
I think making it more generic doesn't hurt since we are calling the thing `Interactive Beam` rather than `ipython Beam` though nothing would work without ipython.

We do support using `Interactive Beam` in ipython shell. It will have fewer problems when users are not using a notebook because there will be no "hidden-state"/"out-of-order-execution"/"deleted-cell". And we support a sampled data-table print when `show` is invoked. Duplicate transform labels in different prompts are allowed in ipython shell. And so on...

Reply:
Should we make the code more generic? Unless this runs in ipython environment it will not do anything even if the environment is interactive. We can either change the code or the name to make them match each other.

Reply:
Changing it to `alter_label_if_ipython` to be more descriptive.